File: app/src/main/java/com/education4all/MathCoachAlg/DataReader.java
package com.education4all.MathCoachAlg;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.StringTokenizer;


public class DataReader {
    public static final String COMPLEXITY_SETTINGS = "ComplexitySettings";
    public static final String ROUND_TIME_SETTINGS = "RoundTimeSettings";
    private static final HashMap<String, Integer> defaultValues = new HashMap<>();

    static {
        defaultValues.put("RoundTime", 1);
        defaultValues.put("DisapRoundTime", -1);
        defaultValues.put("TimerState", 1);
        defaultValues.put("ButtonsPlace", 0);
        defaultValues.put("LayoutState", 0);
        defaultValues.put("Goal", 5);
        defaultValues.put("Theme", -1);
    }

    public static int[] convertIntegers(ArrayList<Integer> integers) {
        int[] ret = new int[integers.size()];
        for (int i = 0; i < ret.length; i++) {
            ret[i] = integers.get(i);
        }
        return ret;
    }

    static private int[] StrToIntArr(String savedString) {
        StringTokenizer st = new StringTokenizer(savedString, ",");
        ArrayList<Integer> savedList = new ArrayList<>();
        int i = 0;
        while (st.hasMoreTokens()) {
            savedList.add(Integer.parseInt(st.nextToken()));
        }
        return convertIntegers(savedList);
    }

    static public int[][] readAllowedTasks(Context p_context) {
        SharedPreferences prefs = p_context.getSharedPreferences(COMPLEXITY_SETTINGS, Context.MODE_PRIVATE);
        String savedString = "";
        int[] Add;
        if (prefs.contains("Add")) {
            savedString = prefs.getString("Add", null);
            Add = StrToIntArr(savedString);
        } else {
            Add = new int[0];
        }
        int[] Sub;
        if (prefs.contains("Sub")) {
            savedString = prefs.getString("Sub", null);
            Sub = StrToIntArr(savedString);
        } else {
            Sub = new int[0];
        }
        int[] Mul;
        if (prefs.contains("Mul")) {
            savedString = prefs.getString("Mul", null);
            Mul = StrToIntArr(savedString);
        } else {
            Mul = new int[0];
        }
        int[] Div;
        if (prefs.contains("Div")) {
            savedString = prefs.getString("Div", null);
            Div = StrToIntArr(savedString);
        } else {
            Div = new int[0];
        }


        return new int[][]{Add, Sub, Mul, Div};
    }

    static public boolean checkComplexity(int p_action, int p_complexity, Context p_context) {
        int[][] l_allowedTasks = DataReader.readAllowedTasks(p_context);
        for (int i = 0; i < l_allowedTasks[p_action].length; ++i) {
            if (l_allowedTasks[p_action][i] == p_complexity) {
                return true;
            }
        }
        return false;
    }

    public static void SaveValue(int value, String name, Context p_context) {
        SharedPreferences.Editor editor = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit();
        editor.putInt(name, value);
        editor.commit();
    }

    static public int GetValue(String name, Context p_context) {
        SharedPreferences prefs = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE);
        return prefs.getInt(name, defaultValues.get(name));
    }

    public static void SaveQueue(String json, Context p_context) {
        SharedPreferences.Editor editor = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit();
        editor.putString("Queue", json);
        editor.commit();
    }

    static public String GetQueue(Context p_context) {
        SharedPreferences prefs = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE);
        return prefs.getString("Queue", "");
    }

    public static void SaveStat(String json, Context p_context) {
        SharedPreferences.Editor editor = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit();
        editor.putString("Stat", json);
        editor.commit();
    }

    static public String GetStat(Context p_context) {
        SharedPreferences prefs = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE);
        return prefs.getString("Stat", "");
    }

}

File: app/src/main/java/com/education4all/MathCoachAlg/DataReader.kt
package com.education4all.mathCoachAlg

import android.content.*
import java.util.*

object DataReader {
    const val COMPLEXITY_SETTINGS = "ComplexitySettings"
    const val ROUND_TIME_SETTINGS = "RoundTimeSettings"
    const val ROUND_TIME = "RoundTime"
    const val DISAP_ROUND_TIME = "DisapRoundTime"
    const val TIMER_STATE = "TimerState"
    const val BUTTONS_PLACE = "ButtonsPlace"
    const val LAYOUT_STATE = "LayoutState"
    const val GOAL = "Goal"
    const val THEME = "Theme"
    const val REMINDER = "Reminder"
    const val REMINDER_TIME = "ReminderTime"
    const val QUEUE = "QueueEnabled"
    const val SIGNEDINTOFIREBASE = "SignedInToFirebase"
    private val defaultInts = HashMap<String?, Int>()
    private val defaultStrings = HashMap<String?, String>()
    private val defaultBooleans = HashMap<String?, Boolean>()
    fun convertIntegers(integers: ArrayList<Int>): IntArray {
        val ret = IntArray(integers.size)
        for (i in ret.indices) {
            ret[i] = integers[i]
        }
        return ret
    }

    private fun StrToIntArr(savedString: String?): IntArray {
        val st = StringTokenizer(savedString, ",")
        val savedList = ArrayList<Int>()
        val i = 0
        while (st.hasMoreTokens()) {
            savedList.add(st.nextToken().toInt())
        }
        return convertIntegers(savedList)
    }

    fun readAllowedTasks(p_context: Context?): Array<IntArray> {
        val prefs = p_context!!.getSharedPreferences(COMPLEXITY_SETTINGS, Context.MODE_PRIVATE)
        var savedString: String? = ""
        val Add: IntArray
        if (prefs.contains("Add")) {
            savedString = prefs.getString("Add", null)
            Add = StrToIntArr(savedString)
        } else {
            Add = IntArray(0)
        }
        val Sub: IntArray
        if (prefs.contains("Sub")) {
            savedString = prefs.getString("Sub", null)
            Sub = StrToIntArr(savedString)
        } else {
            Sub = IntArray(0)
        }
        val Mul: IntArray
        if (prefs.contains("Mul")) {
            savedString = prefs.getString("Mul", null)
            Mul = StrToIntArr(savedString)
        } else {
            Mul = IntArray(0)
        }
        val Div: IntArray
        if (prefs.contains("Div")) {
            savedString = prefs.getString("Div", null)
            Div = StrToIntArr(savedString)
        } else {
            Div = IntArray(0)
        }
        return arrayOf(Add, Sub, Mul, Div)
    }

    fun checkComplexity(p_action: Int, p_complexity: Int, p_context: Context?): Boolean {
        val l_allowedTasks = readAllowedTasks(p_context)
        for (i in 0 until l_allowedTasks[p_action].length) if (l_allowedTasks[p_action][i] == p_complexity) return true
        return false
    }

    fun SaveInt(value: Int, name: String?, p_context: Context?) {
        val editor =
            p_context!!.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit()
        editor.putInt(name, value)
        editor.apply()
    }

    fun GetInt(name: String?, p_context: Context?): Int {
        return if (!defaultInts.containsKey(name)) throw IllegalArgumentException() else {
            val prefs = p_context!!.getSharedPreferences(
                ROUND_TIME_SETTINGS,
                Context.MODE_PRIVATE
            )
            prefs.getInt(name, defaultInts[name]!!)
        }
    }

    fun SaveString(value: String?, name: String?, p_context: Context?) {
        val editor =
            p_context!!.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit()
        editor.putString(name, value)
        editor.apply()
    }

    fun GetString(name: String?, p_context: Context?): String? {
        return if (!defaultStrings.containsKey(name)) throw IllegalArgumentException() else {
            val prefs = p_context!!.getSharedPreferences(
                ROUND_TIME_SETTINGS,
                Context.MODE_PRIVATE
            )
            prefs.getString(name, defaultStrings[name])
        }
    }

    fun SaveBoolean(value: Boolean?, name: String?, p_context: Context?) {
        val editor =
            p_context!!.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit()
        editor.putBoolean(name, value!!)
        editor.apply()
    }

    fun GetBoolean(name: String?, p_context: Context?): Boolean {
        return if (!defaultBooleans.containsKey(name)) throw IllegalArgumentException() else {
            val prefs = p_context!!.getSharedPreferences(
                ROUND_TIME_SETTINGS,
                Context.MODE_PRIVATE
            )
            prefs.getBoolean(name, defaultBooleans[name]!!)
        }
    }

    fun SaveQueue(json: String?, p_context: Context) {
        val editor =
            p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit()
        editor.putString("Queue", json)
        editor.apply()
    }

    fun GetQueue(p_context: Context): String? {
        val prefs = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE)
        return prefs.getString("Queue", "")
    }

    fun SaveStat(json: String?, p_context: Context) {
        val editor =
            p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE).edit()
        editor.putString("Stat", json)
        editor.apply()
    }

    fun GetStat(p_context: Context): String? {
        val prefs = p_context.getSharedPreferences(ROUND_TIME_SETTINGS, Context.MODE_PRIVATE)
        return prefs.getString("Stat", "")
    }

    init {
        //ints
        defaultInts[ROUND_TIME] = 1
        defaultInts[DISAP_ROUND_TIME] = -1
        defaultInts[GOAL] = 5

        //enums
        defaultInts[TIMER_STATE] = 1 // continuous
        defaultInts[BUTTONS_PLACE] = 0 // right
        defaultInts[LAYOUT_STATE] = 0 // 789
        defaultInts[THEME] = -1 // dark

        //booleans
        defaultBooleans[REMINDER] = false
        defaultBooleans[QUEUE] = false
        defaultBooleans[SIGNEDINTOFIREBASE] = false

        //strings
        defaultStrings[REMINDER_TIME] = "18:00"
    }
}